### ISSUE TITLE
functional-tester: kill leader returns different hash

### ISSUE THREAD

**Issue by gyuho** [coreos/etcd#5220]
```
2016-04-28 05:17:08.889242 I | etcd-tester: [round#4686 case#3] start failure kill leader member
2016-04-28 05:17:08.889285 I | etcd-tester: [round#4686 case#3] start injecting failure...
2016-04-28 05:17:08.919163 I | etcd-tester: [round#4686 case#3] injected failure
2016-04-28 05:17:08.919202 I | etcd-tester: [round#4686 case#3] start recovering failure...
2016-04-28 05:17:11.204553 I | etcd-tester: [round#4686 case#3] recovered failure
2016-04-28 05:17:11.204587 I | etcd-tester: [round#4686 case#3] canceling the stressers...
2016-04-28 05:17:11.206395 I | etcd-tester: [round#4686 case#3] canceled stressers
2016-04-28 05:17:11.206410 I | etcd-tester: [round#4686 case#3] checking current revisions...
2016-04-28 05:17:12.377582 I | etcd-tester: [round#4686 case#3] all members are consistent with current revisions [revisions: map[10.240.0.2:2379:2417063585 10.240.0.3:2379:2417063585 10.240.0.4:2379:2417063585]]
2016-04-28 05:17:12.377629 I | etcd-tester: [round#4686 case#3] checking current storage hashes...
2016-04-28 05:17:12.377646 I | etcd-tester: [round#4686 case#3] checking current storage hashes failed [hashes: map[10.240.0.3:2379:1719633027 10.240.0.4:2379:1280481596 10.240.0.2:2379:1719633027]]
2016-04-28 05:17:12.377676 I | etcd-tester: [round#4686 case#3] cleaning up...

```

more logs coming...

**Comment by xiang90** [comment]
Let's finish the analysis offline first. Or the issues might become messy.

**Comment by gyuho** [comment]
Sure. Will figure out what happen first.

**Comment by gyuho** [comment]
```
2016-04-28 05:17:08.889242 I | etcd-tester: [round#4686 case#3] start failure kill leader member
2016-04-28 05:17:08.889285 I | etcd-tester: [round#4686 case#3] start injecting failure...
2016-04-28 05:17:08.919163 I | etcd-tester: [round#4686 case#3] injected failure
2016-04-28 05:17:08.919202 I | etcd-tester: [round#4686 case#3] start recovering failure...
2016-04-28 05:17:11.204553 I | etcd-tester: [round#4686 case#3] recovered failure
2016-04-28 05:17:11.204587 I | etcd-tester: [round#4686 case#3] canceling the stressers...
2016-04-28 05:17:11.206395 I | etcd-tester: [round#4686 case#3] canceled stressers
2016-04-28 05:17:11.206410 I | etcd-tester: [round#4686 case#3] checking current revisions...
2016-04-28 05:17:12.377582 I | etcd-tester: [round#4686 case#3] all members are consistent with current revisions [revisions: map[10.240.0.2:2379:2417063585 10.240.0.3:2379:2417063585 10.240.0.4:2379:2417063585]]
2016-04-28 05:17:12.377629 I | etcd-tester: [round#4686 case#3] checking current storage hashes...
2016-04-28 05:17:12.377646 I | etcd-tester: [round#4686 case#3] checking current storage hashes failed [hashes: map[10.240.0.3:2379:1719633027 10.240.0.4:2379:1280481596 10.240.0.2:2379:1719633027]]
2016-04-28 05:17:12.377676 I | etcd-tester: [round#4686 case#3] cleaning up...
```

So different hash values are:

```
AGENT1 10.240.0.2:2379:1719633027
AGENT2 10.240.0.3:2379:1719633027
AGENT3 10.240.0.4:2379:1280481596
```

If read just the snapshot status, we can see the final hash values are equivalent.


```
# AGENT1
Total Size:       211 MB
Current Revision: 2417063585
Consistent Index: 2417848520
Total Key:        461845
Hash:             1719633027 (667F8483)

# AGENT2
Total Size:       211 MB
Current Revision: 2417063585
Consistent Index: 2417848520
Total Key:        461845
Hash:             1719633027 (667F8483)

# AGENT3
Total Size:       211 MB
Current Revision: 2417063585
Consistent Index: 2417848520
Total Key:        461845
Hash:             1719633027 (667F8483)
```

Then where `1280481596` of AGENT3 does this come from?
I found it here at revision 2417063449 (136 revisions behind).

```
# AGENT3
Total Size:       211 MB
Current Revision: 2417063449
Consistent Index: 2417848384
Total Key:        461709
Hash:             1280481596 (4C52993C)
```

Which means data were consistent, but AGENT3 was just falling behind.

1. We killed the leader (AGENT2) at `2016-04-28 05:17:08.889285`.
2. AGENT1 became the new leader at `2016-04-28 05:17:10.974413`.
3. We recovered it at `2016-04-28 05:17:11.204553`.
4. We canceled stressers at `2016-04-28 05:17:11.206395`.
5. We got hash, revisions at `2016-04-28 05:17:11.206410`.

This means we had only **2m between recovery and hash check**, and AGEN3 was
never a leader BEFORE(AGENT2) and AFTER(AGENT1) failure injection.

We might have to wait more like we do https://github.com/coreos/etcd/pull/5164/files.


```
# AGENT1
05:17:09.857965 I | raft: AGENT1 is starting a new election at term 34183
05:17:09.858036 I | raft: AGENT1 became candidate at term 34184
05:17:09.858046 I | raft: AGENT1 received vote from AGENT1 at term 34184
05:17:09.858058 I | raft: AGENT1 [logterm: 34183, index: 2417843738] sent vote request to AGENT2 at term 34184
05:17:09.858074 I | raft: AGENT1 [logterm: 34183, index: 2417843738] sent vote request to AGENT3 at term 34184
05:17:09.858086 I | raft: raft.node: AGENT1 lost leader AGENT2 at term 34184
05:17:09.859340 E | rafthttp: failed to write AGENT2 on stream Message (write tcp 10.240.0.2:2380->10.240.0.3:58872: write: broken pipe)
05:17:09.859357 I | rafthttp: the connection with AGENT2 became inactive
05:17:09.861089 I | raft: AGENT1 received vote from AGENT3 at term 34184
05:17:09.861117 I | raft: AGENT1 [quorum:2] has received 2 votes and 0 vote rejections
05:17:09.861155 I | raft: AGENT1 became leader at term 34184
05:17:09.861195 I | raft: raft.node: AGENT1 elected leader AGENT1 at term 34184

# AGENT2
2016-04-28 05:17:08.940383 I | etcdmain: etcd Version: 2.3.0+git
2016-04-28 05:17:08.940428 I | etcdmain: Git SHA: Not provided (use ./build instead of go build)
2016-04-28 05:17:08.940435 I | etcdmain: Go Version: go1.6.2
2016-04-28 05:17:08.940441 I | etcdmain: Go OS/Arch: linux/amd64
2016-04-28 05:17:08.940451 I | etcdmain: setting maximum number of CPUs to 4, total number of available CPUs is 4
2016-04-28 05:17:08.940484 N | etcdmain: the server is already initialized as member before, starting as etcd member...
2016-04-28 05:17:08.940549 I | etcdmain: listening for peers on http://10.240.0.3:2380
2016-04-28 05:17:08.940577 I | etcdmain: listening for client requests on 10.240.0.3:2379
2016-04-28 05:17:08.955503 W | snap: skipped unexpected non snapshot file 000000007805aea6.snap.db
2016-04-28 05:17:08.955525 W | snap: skipped unexpected non snapshot file 00000000433851a8.snap.db
2016-04-28 05:17:08.974341 I | etcdserver: recovered store from snapshot at index 2417786847
2016-04-28 05:17:08.974369 I | etcdserver: name = etcd-1
2016-04-28 05:17:08.974379 I | etcdserver: data dir = agent.etcd
2016-04-28 05:17:08.974391 I | etcdserver: member dir = agent.etcd/member
2016-04-28 05:17:08.974403 I | etcdserver: heartbeat = 100ms
2016-04-28 05:17:08.974425 I | etcdserver: election = 1000ms
2016-04-28 05:17:08.974440 I | etcdserver: snapshot count = 10000
2016-04-28 05:17:08.974455 I | etcdserver: advertise client URLs = http://10.240.0.3:2379
2016-04-28 05:17:09.212260 I | etcdserver: restarting member AGENT2 in cluster e5a1de03b0c74677 at commit index 2417843666
2016-04-28 05:17:09.213874 I | raft: AGENT2 became follower at term 34183
2016-04-28 05:17:09.213901 I | raft: newRaft AGENT2 [peers: [AGENT1,AGENT2,AGENT3], term: 34183, commit: 2417843666, applied: 2417786847, lastindex: 2417843738, lastterm: 34183]
2016-04-28 05:17:09.230969 I | membership: added member AGENT1 [http://10.240.0.2:2380] to cluster e5a1de03b0c74677 from store
2016-04-28 05:17:09.231007 I | membership: added member AGENT2 [http://10.240.0.3:2380] to cluster e5a1de03b0c74677 from store
2016-04-28 05:17:09.231020 I | membership: added member AGENT3 [http://10.240.0.4:2380] to cluster e5a1de03b0c74677 from store
2016-04-28 05:17:09.231034 I | membership: set the cluster version to 2.3 from store
2016-04-28 05:17:09.240583 I | storage: restore compact to 2416851748
2016-04-28 05:17:10.971354 I | etcdserver: starting server... [version: 2.3.0+git, cluster version: 2.3]
2016-04-28 05:17:10.972042 E | etcdmain: failed to notify systemd for readiness: No socket
2016-04-28 05:17:10.972059 E | etcdmain: forgot to set Type=notify in systemd service file?
2016-04-28 05:17:10.972140 I | v2http: enabled capabilities for version 2.3.0
2016-04-28 05:17:10.972201 I | fileutil: purged file agent.etcd/member/snap/0000000000008583-00000000901b6b19.snap successfully
2016-04-28 05:17:10.972345 I | fileutil: purged file agent.etcd/member/snap/0000000000008583-00000000901b94ca.snap successfully
2016-04-28 05:17:10.972479 I | fileutil: purged file agent.etcd/member/snap/0000000000008583-00000000901bbbdd.snap successfully
2016-04-28 05:17:10.972541 I | rafthttp: the connection with AGENT1 became active
2016-04-28 05:17:10.972675 N | etcdmain: serving insecure client requests on 10.240.0.3:2379, this is strongly discouraged!
2016-04-28 05:17:10.973479 I | rafthttp: the connection with AGENT3 became active
2016-04-28 05:17:10.974109 I | raft: AGENT2 [term: 34183] received a MsgHeartbeat message with higher term from AGENT1 [term: 34184]
2016-04-28 05:17:10.974383 I | raft: AGENT2 became follower at term 34184
2016-04-28 05:17:10.974413 I | raft: raft.node: AGENT2 elected leader AGENT1 at term 34184
2016-04-28 05:17:11.088522 I | etcdserver: start to snapshot (applied: 2417843666, lastsnap: 2417786847)
2016-04-28 05:17:11.105573 I | etcdserver: saved snapshot at index 2417843666
2016-04-28 05:17:11.106008 I | etcdserver: compacted raft log at 2417838666
2016-04-28 05:17:11.174427 I | etcdserver: published {Name:etcd-1 ClientURLs:[http://10.240.0.3:2379]} to cluster e5a1de03b0c74677

# AGENT3
05:17:08.899181 E | rafthttp: failed to write AGENT2 on stream Message (write tcp 10.240.0.4:2380->10.240.0.3:57783: write: broken pipe)
05:17:08.899212 I | rafthttp: the connection with AGENT2 became inactive
05:17:09.859943 I | raft: AGENT3 [term: 34183] received a MsgVote message with higher term from AGENT1 [term: 34184]
05:17:09.859999 I | raft: AGENT3 became follower at term 34184
05:17:09.860023 I | raft: AGENT3 [logterm: 34183, index: 2417843738, vote: 0] voted for AGENT1 [logterm: 34183, index: 2417843738] at term 34184
05:17:09.860042 I | raft: raft.node: AGENT3 lost leader AGENT2 at term 34184
05:17:09.861755 I | raft: raft.node: AGENT3 elected leader AGENT1 at term 34184
05:17:10.974645 I | rafthttp: the connection with AGENT2 became active
```

<br><hr>
#### Solution

We've decided to make `Hash` API more atomic with revisions, so that functional-tester can retry with more accurate revisions (https://github.com/coreos/etcd/pull/5248).
